From 17b96edf816d235648ea8e1fb3ac82f06cabb6f7 Mon Sep 17 00:00:00 2001
From: Hariprasad Kelam <hkelam@marvell.com>
Date: Fri, 26 Mar 2021 10:11:50 +0530
Subject: [PATCH 1390/1921] octeontx2-pf: Read correct number of VFs

In function sriov_disable , while freeing resources number of
vfs is read from pci_num_vf(). But this api returning value as 0
due to pci_disable_sriov internally clears number of vfs.
Fix this by calling pci_disable_sriov after reading pci_num_vf().

Change-Id: Ia453dd607b3acdc5664c6a704fd621d8239102eb
Fixes: 37a6d8b4ab60 ("octeontx2-pf: Notify VFs about link changes")
Signed-off-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48942
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index ad3013b07e90..2646b4f39f4d 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -2891,12 +2891,12 @@ static int otx2_sriov_disable(struct pci_dev *pdev)
 	if (!numvfs)
 		return 0;
 
-	pci_disable_sriov(pdev);
-
 	for (i = 0; i < pci_num_vf(pdev); i++) {
 		cancel_delayed_work_sync(&pf->vf_configs[i].link_event_work);
 		cancel_delayed_work_sync(&pf->vf_configs[i].ptp_info_work);
 	}
+
+	pci_disable_sriov(pdev);
 	kfree(pf->vf_configs);
 
 	otx2_disable_flr_me_intr(pf);
-- 
2.31.1

